### DISCUSSION TITLE
[BUG] StatusBarColor with AppThemeBinding doesn't reflect changes of light/dark mode on the fly

### DISCUSSION THREAD

**Discussion by RsZoli** [CommunityToolkit/Maui#2392] [New Feature Discussions]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [x] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

When setting the StatusBarColor with AppThemeBinding,  changes of light/dark mode doesn't get reflected on the fly!

The page needs to be navigated to once more for the changes to take effect!

https://github.com/user-attachments/assets/69056dc6-99fd-4c33-87bb-684e3e391d72

### Expected Behavior

Changes should take effect on the fly!

### Steps To Reproduce

1. Open and run solution from reproduction repository.
2. Click the button titled "StatusbarBehavior repro 3" and observe the bug.

### Link to public reproduction project repository

https://github.com/RsZoli/GitHubRepros

### Environment

- .NET MAUI CommunityToolkit: 9.1.0
- OS: Windows 11 22631.4317
- .NET MAUI:
android                    34.0.138/8.0.100       SDK 8.0.400, VS 17.11.35327.3
ios                        18.0.8303/8.0.100      SDK 8.0.400, VS 17.11.35327.3
maui-windows               8.0.82/8.0.100         SDK 8.0.400, VS 17.11.35327.3

**Comment by jakob-bmd** [comment]
I created another repo based on @jfversluis [MauiToolkitStatusBarBehaviorSample](https://github.com/jfversluis/MauiToolkitStatusBarBehaviorSample):

https://github.com/jakob-bmd/MauiToolkitStatusBarBehaviorAppThemeBindingBug/tree/main 

**Steps To Reproduce**
GIVEN: user is in mainpage of the app
WHEN: changing the android OS Theme to (e.g.) black and navigating back to the app
THEN: the statusbar stays white

![Image](https://github.com/user-attachments/assets/e3c1240f-b7b5-426e-9f55-3fd2c3ab5560)

**Expected Behavior**
the color of the statusbar should change immediately with the OS theme change

**Additional Info**
The statusbar color changes correctly, when hotreload takes place

**Environment**
- .NET MAUI CommunityToolkit: 9.1.1
- MauiVersion 8.0.93
- Android 33

**Comment by vniehues** [comment]
My current (crude) workaround is subscribing to the RequestedThemeChanged event and manually setting the colors and styles but I feel like this functionality should be in the Behavior itself. 

```
        Application.Current.RequestedThemeChanged += (sender, args) =>
        {
            switch (args.RequestedTheme)
            {
                case AppTheme.Dark:
                    StatusBar.SetColor(Application.Current.Resources["OffBlack"] as Color);
                    StatusBar.SetStyle(StatusBarStyle.LightContent);
                    break;
                case AppTheme.Light:
                case AppTheme.Unspecified:
                    StatusBar.SetColor(Colors.White);
                    StatusBar.SetStyle(StatusBarStyle.DarkContent);
                    break;
            }
        };
```

**Reply by RsZoli** [reply]
> My current (crude) workaround is subscribing to the RequestedThemeChanged event and manually setting the colors and styles but I feel like this functionality should be in the Behavior itself.
> 
> ```
>         Application.Current.RequestedThemeChanged += (sender, args) =>
>         {
>             switch (args.RequestedTheme)
>             {
>                 case AppTheme.Dark:
>                     StatusBar.SetColor(Application.Current.Resources["OffBlack"] as Color);
>                     StatusBar.SetStyle(StatusBarStyle.LightContent);
>                     break;
>                 case AppTheme.Light:
>                 case AppTheme.Unspecified:
>                     StatusBar.SetColor(Colors.White);
>                     StatusBar.SetStyle(StatusBarStyle.DarkContent);
>                     break;
>             }
>         };
> ```

@vniehues  hey! it seems that this solution doesnt work anymore on android 16! :( do you have an update, maybe?

**Comment by vhugogarcia** [comment]
Hi @vniehues, thanks so much for reporting this issue!

It seems that this isn't a bug, but rather a topic for discussion around integrating a feature to detect when a requested theme has changed. Because of this, I’m moving this to a discussion. If it gains enough interest and votes, it can be considered for future implementation.

**Comment by stephenquan** [comment]
Another approach is to expose `StatusBarColor` and `StatusBarStyle` as `BindableProperties` on your `ContentPage`. The implementation uses `CoerceValue` to invoke the corresponding CommunityToolkit.Maui methods directly.

```c#
// MainPage.xaml
public partial class MainPage : ContentPage
{
    [BindableProperty(CoerceValueMethodName = nameof(CoerceStatusBarColor))]
    public partial Color? StatusBarColor { get; set; }
    static object CoerceStatusBarColor(BindableObject bindable, object value)
        => PlatformInvokeWithCoerceValue<Color?>(value, CommunityToolkit.Maui.Core.Platform.StatusBar.SetColor);
    [BindableProperty(CoerceValueMethodName = nameof(CoerceStatusBarStyle))]
    public partial StatusBarStyle StatusBarStyle { get; set; }
    static object CoerceStatusBarStyle(BindableObject bindable, object value)
        => PlatformInvokeWithCoerceValue<StatusBarStyle>(value, CommunityToolkit.Maui.Core.Platform.StatusBar.SetStyle);
    static object PlatformInvokeWithCoerceValue<T>(object value, Action<T> action)
    {
#if ANDROID || IOS
        action((T)value);
#endif
        return value;
    }
    // ...
```

Then, you can happily use AppThemeBinding in XAML:

```xaml
<ContentPage
    x:Class="Maui.StatusBarDemo.MainPage"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    StatusBarColor="{AppThemeBinding Light=#c0c, Dark=#303}"
    StatusBarStyle="{AppThemeBinding Light=DarkContent, Dark=LightContent}">
    <!-- ... -->
```

**Comment by cklumph** [comment]
This bug is still around in CommunityToolkit 14.0.0 and MAUI 10.0.40. Tested using the toolkit:StatusBarBehavior<...> functionality on iOS 26. Status bar doesn't change color until you navigate to a new page. I haven't tried the workarounds listed on this page, but they probably still apply.

**Comment by stephenquan** [comment]
I did some troubleshooting and found that when the app is not focused and you change the app theme, the AppThemeBinding should react to those changes, but they do not propagate to BindableProperties on Behaviors. I could not find a workaround for Behaviors. However, when I connected the AppThemeBinding to BindableProperties on an attached property, it worked correctly.

GitHub repo: https://github.com/stephenquan/MauiStatusBarTest

```c#
[AttachedBindableProperty<Color>("StatusBarColor", CoerceValueMethodName = nameof(CoerceStatusBarColor))]
[AttachedBindableProperty<StatusBarStyle>("StatusBarStyle", CoerceValueMethodName = nameof(CoerceStatusBarStyle))]
public static partial class StatusBarExtras
{
    static object CoerceStatusBarColor(BindableObject bindable, object value)
        => PlatformInvokeWithCoerceValue<Color?>(value, CommunityToolkit.Maui.Core.Platform.StatusBar.SetColor);
    static object CoerceStatusBarStyle(BindableObject bindable, object value)
        => PlatformInvokeWithCoerceValue<StatusBarStyle>(value, CommunityToolkit.Maui.Core.Platform.StatusBar.SetStyle);
    static object PlatformInvokeWithCoerceValue<T>(object value, Action<T> action)
    {
#if ANDROID || IOS
        action((T)value);
#endif
        return value;
    }
}
```

```xaml
<ContentPage
    x:Class="MauiStatusBarTest.MainPage"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:local="clr-namespace:MauiStatusBarTest"
    xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
    local:StatusBarExtras.StatusBarColor="{AppThemeBinding Light=#c0c, Dark=#303}"
    local:StatusBarExtras.StatusBarStyle="{AppThemeBinding Light=DarkContent, Dark=LightContent}">
    <!--
    <ContentPage.Behaviors>
        <toolkit:StatusBarBehavior
            StatusBarColor="{AppThemeBinding Light=#c0c, Dark=#303}"
            StatusBarStyle="{AppThemeBinding Light=DarkContent, Dark=LightContent}" />
    </ContentPage.Behaviors>
    -->
```

<img width="270" height="585" alt="image" src="https://github.com/user-attachments/assets/289086ee-fb1f-482d-ac86-99108a99ac7a" />

<img width="270" height="585" alt="image" src="https://github.com/user-attachments/assets/65a5dd85-df14-4232-adb2-a62b6bad49ae" />

**Comment by AncientLust** [comment]
This is definitely a bug. `StatusBarStyle` doesn't reflect changes on `AppThemeBinding`. The same applies if `AppThemeBinding` replaced  by `DynamicResource`. Nothing happens on theme change. `StatusBarColor` at least works on the first assignment. `StatusBarStyle` doesn't react even on the first assignment if `DynamicResource` used.

CommunityToolkit.Maui.Core 14.0.1

Status bar controls are still complete mess.
